src/Features/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/DotnetCliHelper.cs

```diff
@@ -0,0 +1,172 @@
+        {
+            return dotnetRootUserExecutable;
+        }
+
+        // Neither an option or env var was provided, find the dotnet path using what is currently on the path.
+        _logger.LogInformation("Using dotnet executable configured on the PATH");
+        return "dotnet";
+
+        static bool TryGetDotnetExecutableFromFolder(string? dotnetFolder, string optionName, ILogger logger, [NotNullWhen(true)] out string? dotnetExecutablePath)
+        {
+            if (string.IsNullOrEmpty(dotnetFolder))
+            {
+                dotnetExecutablePath = null;
+                return false;
+            }
+
+            var dotnetExecutableName = $"dotnet{(RuntimeInformation.IsOSPlatform(OSPlatform.Windows) ? ".exe" : string.Empty)}";
+            var executablePath = Path.Combine(dotnetFolder, dotnetExecutableName);
+            if (File.Exists(executablePath))
+            {
+                logger.LogInformation($"Found dotnet executable at {executablePath} using {optionName}");
+                dotnetExecutablePath = executablePath;
+                return true;
+            }
+            else
+            {
+                logger.LogInformation($"The {optionName} option {dotnetFolder} does not contain a valid dotnet executable");
+                dotnetExecutablePath = null;
+                return false;
+            }
+        }
+    }
+
+    /// <summary>
+    /// The folder the dotnet executable is in could contain multiple SDK paths.
+    /// In order to figure out which one is the right one, we need to run dotnet --info
+    /// which will output the correct SDK path.
+    /// </summary>
+    private async Task<string> GetDotnetSdkFolderFromDotnetExecutableAsync(CancellationToken cancellationToken)
+    {
+        using var process = Run("--info", workingDirectory: null, shouldLocalizeOutput: false);
+
+        string? dotnetSdkFolderPath = null;
+        process.OutputDataReceived += (_, e) =>
+        {
+            if (!string.IsNullOrEmpty(e.Data))
+            {
+                var line = e.Data.Trim();
+                if (line.StartsWith("Base Path", StringComparison.OrdinalIgnoreCase))
+                {
+                    dotnetSdkFolderPath = e.Data.Split(':', count: 2)[1].Trim();
+                }
+            }
+        };
+
+        var errorOutput = new StringBuilder();
+        process.ErrorDataReceived += (_, e) => errorOutput.AppendLine(e.Data);
+
+        process.BeginOutputReadLine();
+        process.BeginErrorReadLine();
+        await process.WaitForExitAsync(cancellationToken);
+        if (process.ExitCode != 0 || dotnetSdkFolderPath == null)
+        {
+            _logger.LogError(errorOutput.ToString());
+            throw new InvalidOperationException("Failed to get dotnet SDK folder from dotnet --info");
+        }
+
+        return dotnetSdkFolderPath;
+    }
+
+    public Process Run(string arguments, string? workingDirectory, bool shouldLocalizeOutput = true)
```

Comment:
woudl prefer this parametre not be optional.  the we can check if callers are thinking about this case.